rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,310 @@
+* Set 2: `PluggableDevicePlatform` related classes 
+   * class `PluggableDevicePlatform`: PluggableDevice-specific platform, its platform name is "PluggableDevice", it contains a C struct: SE_Platform* platform_ which is its internal implementation and as the C interface registered by device plugin.
+   * class `PluggableDeviceExecutor`: PluggableDevice-platform implementation of the platform-agnostic StreamExecutorInterface, it contains C structs: SE_StreamExecutor* executor_ and SE_Device* device_ whose member can be accessed in both TensorFlow proper and device plugins.
+   * class `PluggableDeviceStream`: wraps a StreamHandle in order to satisfy the platform-independent StreamInterface. It returns SE_Stream which is treated as an opaque type to TensorFlow,  whose structure is created by the device plugin.  
+   * class `PluggableDeviceTimer`: wraps an opaque handle: SE_Timer to satisfy the platform-independent TimerInterface.
+   * class `PluggableDeviceEvent`: wraps an opaque handle: SE_Event to satisfy the platform-independent EventInterface.
+
+**TensorFlow Plugin**
+
+Plugin authors need to provide those C functions implementation defined in StreamExecutor C API . 
+*  `SE_StreamExecutor` is defined as struct in the C API, both sides(TensorFlow proper and plugins) can access its members. Plugin creates the SE_StreamExecutor and registers its C API implementations to the SE_StreamExecutor.  
+```cpp
+   SE_StreamExecutor* create_stream_executor(TF_Status* status) {
+     SE_StreamExecutor* se_nfs = new SE_StreamExecutor{ SE_STREAMEXECUTOR_STRUCT_SIZE };
+     se->memcpy_from_host = my_device_memory_from_host_function;
+     se->allocate = my_allocate_function;
+     …
+   }//Init device
+```
+* `SE_Device` is defined as struct in the C API, both sides(TensorFlow proper and plugins) can access its members. Plugin creates the SE_Device and fills its device opaque handle and device name to the SE_Device.
+```cpp
+  SE_Device* create_device(SE_Options* options, TF_Status* status) {
+    SE_Device* se = new SE_Device( SE_DEVICE_STRUCT_SIZE );
+    se->device_handle = get_my_device_handle();
+    ...
+    return se;
+  }
+```
+* `SE_Stream` is defined in plugin and treated as an opaque struct in TensorFlow proper. 
+```cpp
+  void create_stream(SE_Device* executor, SE_Stream* stream, TF_Status*) {
+    *stream = new SE_Stream_st();
+    (*stream)->stream_handle = create_my_stream_handle(executor);
+    ..
+  }
+```
+
+### PluggableDevice kernel registration
+
+This RFC shows an example of registering kernels for PluggableDevice. Kernel and op registration and implementation API is addressed in a separate [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md). 
+
+To avoid kernel registration conflict with existing GPU(CUDA) kernels, the backend device_type for kernel registration should be seperated from the front-end visible device type ("GPU"). Two Options:  
```

Comment:
@annarev  do you have any suggestion here? I list two options to resolve  the potential kernel registration conflict with existing CUDA kernels when the device type is "GPU", seems both options need StreamExecutor C API/Kernel Registration C API support.Thanks

Reply:
I think I like Option 1 a bit more since it allows backend name to be more descriptive.

There is also another option: specify kernel registration [priority](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.h;l=111). Then, two kernels can be registered with same type.

I think there is a more important question to answer though. Both Option 1 and Option 2 have "GPU" specified as front-end name. So, an important question is: suppose user specifies "GPU" device, which device would be selected? May be we need to provide some way for user to specify that GPU==backend name.

Reply:
@annarev  I thought that only One GPU device will be registered successfully? Since multiple GPU device registration with same device name and priority will got conflict.  Specifying the kernel registration priority will help the new GPU kernel register with the same device type as CUDA Kernels(GPU type), however, there are still lots of code that are specific for CUDA, so  we may still need a new backend device type for those code within GPU device type scope that are specific for CUDA.

Reply:
I was also originally thinking of having one device registered per type. At the same time, I remember a question in StreamExecutor C API design review if there would be a way to use two different GPU types. I don't know if we want to support it. I guess we can discuss more in the design review on Thursday.

Sounds good about keeping the extra name.